src/common/buf.c

```diff
@@ -69,37 +65,38 @@ buf_expand_shell_variables(struct buf *s)
 			buf_add_char(&new, s->buf[i]);
 		}
 	}
-	free(environment_variable.buf);
-	free(s->buf);
-	s->buf = new.buf;
-	s->len = new.len;
-	s->alloc = new.alloc;
+	buf_reset(&environment_variable);
+	buf_move(s, &new);
 }
 
-void
-buf_init(struct buf *s)
+static void
+buf_expand(struct buf *s, int new_alloc)
 {
-	/* we can't assert(!s->buf) here because struct may be uninitialized */
-
-	s->alloc = 256;
-	s->buf = xmalloc(s->alloc);
-	s->buf[0] = '\0';
-	s->len = 0;
+	if (new_alloc <= s->alloc) {
+		return;
+	}
+	new_alloc = MAX(new_alloc, 256);
+	new_alloc = MAX(new_alloc, s->alloc * 3 / 2);
+	if (s->alloc) {
+		assert(s->buf);
+		s->buf = xrealloc(s->buf, new_alloc);
+	} else {
+		assert(!s->len);
+		s->buf = xmalloc(new_alloc);
+		s->buf[0] = '\0';
+	}
+	s->alloc = new_alloc;
 }
 
 void
 buf_add(struct buf *s, const char *data)
 {
-	assert(s->buf);
-
 	if (!data || data[0] == '\0') {
 		return;
 	}
 	int len = strlen(data);
-	if (s->alloc <= s->len + len + 1) {
-		s->alloc = s->alloc + len;
-		s->buf = xrealloc(s->buf, s->alloc);
-	}
+	buf_expand(s, s->len + len + 1);
+	assert(s->alloc);  // dumb but makes gcc static analyzer happy
```

Comment:
I wonder why this is necessary here but not in the `buf_add_char()` case. Maybe the analyzer treats `memcpy()` different to direct assignment. Also all of this are ints so technically `s->alloc` (and `s->len`) could be negative due to overflow but it doesn't seem to bother the analyzer in `buf_add_char()`.

Reply:
I think it's because of the `buf_clear` + `buf_add` sequence used in `buf_expand_shell_variables`.

Modifying the first `if` in `buf_expand` might also prove to the analyzer that `s->buf` is allocated after return:
```
if (s->alloc && new_alloc <= s->alloc) {
	return;
}
```

Reply:
I went with this approach since it seems more logical to me, and added a longer comment.

Reply:
Happy with this.